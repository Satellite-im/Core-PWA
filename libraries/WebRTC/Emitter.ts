/**
 * @class Emitter
 * @description The emitter class provides an interface for
 * listening and emitting custom events in a strongly typed way
 */
export default class Emitter<
  Listeners extends { [key in keyof Listeners]: (...args: any[]) => any },
> {
  private _events: { [key in keyof Listeners]?: Array<Listeners[key]> } = {}

  /**
   * @method on
   * @description Bind event listeners
   * @param event Name of the event to listen to
   * @param listener to call on any Event
   * @example Emitter.on('EVENT_NAME', (parameter: ParameterType) => {})
   */
  on<T extends keyof Listeners>(event: T, listener: Listeners[T]) {
    if (!this._events[event]) {
      this._events[event] = []
    }

    this._events[event]!.push(listener)
  }

  /**
   * @method off
   * @description Removes an event listener from the listener box
   * @param event Name of the event to unsubscribe from
   * @param listener Listener function to remove
   * @example Emitter.off("EVENT_NAME", (parameter: ParameterType) => {})
   */
  off<T extends keyof Listeners>(event: T, listenerToRemove: Listeners[T]) {
    if (!this._events[event]) {
      throw new Error(
        `Can't remove a listener. Event "${event}" doesn't exits.`,
      )
    }

    const filteredListeners = this._events[event]!.filter(
      (listener: Listeners[T]) => listener !== listenerToRemove,
    )

    this._events[event] = filteredListeners
  }

  /**
   * @method emit
   * @description Emits an event to all listeners
   * @param event Name of the event to emit
   * @param data Data to provide to listeners
   * @example Emitter.emit('EVENT_NAME', {eventParam, otherParam})
   */
  protected emit<T extends keyof Listeners>(
    event: T,
    ...[data]: Parameters<Listeners[T]>
  ) {
    if (!this._events[event]) {
      return
    }

    this._events[event]!.forEach((cb) => cb(data))
  }
}
