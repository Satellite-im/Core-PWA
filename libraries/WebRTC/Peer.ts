import SimplePeer from 'simple-peer'

export default class Peer extends SimplePeer {
  constructor(groupId: string, did: string, options: any) {
    super(options)
    this.groupId = groupId
    this.did = did
  }
}
