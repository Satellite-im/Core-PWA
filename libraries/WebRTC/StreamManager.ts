import { Call } from './Call'

export default class StreamManager {

  constructor() {
  }

}
