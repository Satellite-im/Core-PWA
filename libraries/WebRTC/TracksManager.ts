export class TracksManager {
  tracks: { [id: string]: MediaStreamTrack }

  constructor() {
    this.tracks = {}
  }

  addTrack(track: MediaStreamTrack) {
    if (this.tracks[track.id]) {
      console.warn(
        `Track already exists ${track.id} ${track.kind}`,
        this.tracks,
      )
      return
    }

    this.tracks[track.id] = track
  }

  removeTrack(id: string) {
    delete this.tracks[id]
  }

  getTrack(id: string): MediaStreamTrack | undefined {
    return this.tracks[id]
  }

  removeAllTracks() {
    this.tracks = {}
  }
}
