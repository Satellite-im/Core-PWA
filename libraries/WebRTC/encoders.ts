import * as t from 'io-ts'

export const wireBaseMessage = t.type({
  type: t.string,
  payload: t.unknown,
  sentAt: t.number,
})

export const wireIdentificationMessage = t.intersection([
  wireBaseMessage,
  t.type({
    type: t.literal('IDENTIFICATION'),
    payload: t.type({
      peerId: t.string,
    }),
  }),
])

export const wireDataMessage = t.intersection([
  wireBaseMessage,
  t.type({
    type: t.literal('DATA'),
  }),
])

export const wireSignalMessage = t.intersection([
  wireBaseMessage,
  t.type({
    type: t.literal('SIGNAL'),
    payload: t.type({
      peerId: t.string,
      data: t.type({
        type: t.string,
        transceiverRequest: t.type({
          kind: t.string,
          init: t.type({
            sdp: t.string,
          }),
        }),
        renegotiate: t.boolean,
        candidate: t.unknown,
      }),
    }),
  }),
])

export const wireRefuseConnectionMessage = t.intersection([
  wireBaseMessage,
  t.type({
    type: t.literal('REFUSE'),
    payload: t.type({
      peerId: t.string,
    }),
  }),
])

export const KeyboardStates = t.union([
  t.literal('TYPING'),
  t.literal('NOT_TYPING'),
])

export const wireKeyboardState = t.intersection([
  wireBaseMessage,
  t.type({
    type: t.literal('TYPING_STATE'),
    payload: t.type({
      state: KeyboardStates,
    }),
  }),
])
