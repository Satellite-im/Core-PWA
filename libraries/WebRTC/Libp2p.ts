import { Connection, MuxedStream, create } from 'libp2p'
// @ts-ignore
import Mplex from 'libp2p-mplex'
// @ts-ignore
import secio from 'libp2p-secio'
import { NOISE } from 'libp2p-noise'
// @ts-ignore
import WStar from 'libp2p-webrtc-star'
import { SignalData } from 'simple-peer'
import LibP2PCrypto, { keys } from 'libp2p-crypto'
import type Libp2p from 'libp2p/dist/src/index'
import PeerId, {
  createFromB58String,
  createFromPrivKey,
  createFromPubKey,
} from 'peer-id'
import { pipe } from 'it-pipe'
import { Keypair } from '@solana/web3.js'
import Emitter from './Emitter'
import { WireMessage } from './types'
import type { CallPeerDescriptor } from './Call'
import logger from '~/plugins/local/logger'
import { ConversationParticipant } from '~/store/conversation/types'

enum P2PProtocols {
  COMMUNICATION_BUS = '/sattest/chat/1.0.0',
}

interface P2PListeners {
  'peer:discovery': (data: { peerId: PeerId }) => void
  'peer:connect': (data: { peerId: PeerId }) => void
  'peer:disconnect': (data: { peerId: PeerId }) => void
  'peer:onmessage': (data: { peerId: PeerId; message: WireMessage }) => void
  'peer:signal': (data: { peerId: PeerId; payload: SignalData }) => void
  'peer:refuse': (data: { peerId: PeerId; payload: unknown }) => void
  'peer:data': (data: { peerId: PeerId; payload: unknown }) => void
  'peer:typing': (data: { peerId: PeerId; payload: unknown }) => void
  'peer:announce': (data: {
    peerId: PeerId
    payload: ConversationParticipant
  }) => void
  'peer:destroy': (data: { peerId: PeerId; payload: unknown }) => void
  'peer:hangup': (data: { peerId: PeerId; payload: unknown }) => void
  'peer:screenshare': (data: {
    peerId: PeerId
    payload: { streamId: string }
  }) => void
  'peer:mute': (data: {
    peerId: PeerId
    payload: {
      callId: string
      streamId: string
      trackId: string
      kind: string
    }
  }) => void
  'peer:unmute': (data: {
    peerId: PeerId
    payload: {
      callId: string
      streamId: string
      trackId: string
      kind: string
    }
  }) => void
  'peer:call': (data: {
    peerId: PeerId
    payload: {
      callId: string
      peers?: CallPeerDescriptor[]
      signal?: SignalData
    }
  }) => void
}

type P2PMessage = {
  type: string
  payload: unknown
  sentAt: number
}

type SupportedKeyType = keyof typeof LibP2PCrypto.keys.supportedKeys

export interface PrivateKeyInfo {
  privateKey: Uint8Array
  type: SupportedKeyType
}

export interface PublicKeyInfo {
  publicKey: Uint8Array
  type: SupportedKeyType
}

interface InitializationOptions {
  publickey: PublicKeyInfo
}

export class Peer2Peer extends Emitter<P2PListeners> {
  private _node?: Libp2p
  private _peerId?: PeerId
  static instance: Peer2Peer // eslint-disable-line no-use-before-define

  public static getInstance(): Peer2Peer {
    if (!Peer2Peer.instance) {
      Peer2Peer.instance = new Peer2Peer()
    }
    return Peer2Peer.instance
  }

  get peerId() {
    return this._peerId
  }

  get id() {
    return this._peerId?.toB58String()
  }

  async init(opts?: InitializationOptions) {
    if (opts) {
      this._peerId = await this._getPeerIdByType(opts.publickey)
    }

    this._node = await create({
      modules: {
        transport: [WStar],
        streamMuxer: [Mplex],
        connEncryption: [NOISE, secio],
      },
      addresses: {
        listen: [
          '/dns4/wrtc-star1.par.dwebops.pub/tcp/443/wss/p2p-webrtc-star',
          '/dns4/wrtc-star2.sjc.dwebops.pub/tcp/443/wss/p2p-webrtc-star',
        ],
      },
      config: {
        peerDiscovery: {
          mdns: {
            enabled: true,
          },
          webRTCStar: {
            enabled: true,
          },
        },
        pubsub: {
          enabled: true,
          canRelayMessage: true,
        },
      },
      peerId: this.peerId,
      connectionManager: {
        autoDial: false, // Auto connect to discovered peers (limited by ConnectionManager minConnections)
        // The `tag` property will be searched when creating the instance of your Peer Discovery service.
        // The associated object, will be passed to the service when it is instantiated.
      },
    })

    this._bindListeners()
    this._bindProtocols()
  }

  private async _getPeerIdByType(publicKeyInfo: PublicKeyInfo) {
    switch (publicKeyInfo.type) {
      case 'ed25519': {
        window.console.log('publicKeyInfo', publicKeyInfo)
        const publicKey =
          await keys.supportedKeys.ed25519.unmarshalEd25519PublicKey(
            publicKeyInfo.publicKey,
          )

        return createFromPubKey(publicKey.bytes)
      }

      default: {
        break
      }
    }
  }

  start() {
    if (!this._node) return

    this._node.start()
  }

  stop() {
    if (!this._node) return

    this._node.stop()
  }

  async connect(peerId: PeerId) {
    return this._dial(peerId)
  }

  get node() {
    return this._node
  }

  private _bindListeners() {
    if (!this._node) return

    this._node.on('peer:discovery', this._onPeerDiscovery.bind(this))
    this._node.connectionManager.on(
      'peer:connect',
      this._onPeerConnect.bind(this),
    )
    this._node.connectionManager.on(
      'peer:disconnect',
      this._onPeerDisconnect.bind(this),
    )
  }

  private _bindProtocols() {
    if (!this._node) return

    this._node.handle(
      P2PProtocols.COMMUNICATION_BUS,
      this._messageHandler.bind(this),
    )
  }

  private _onPeerDiscovery(peerId: PeerId) {
    this.emit('peer:discovery', { peerId })
  }

  private async _onPeerConnect(connection: Connection) {
    this.emit('peer:connect', { peerId: connection.remotePeer })
  }

  private _onPeerDisconnect(connection: Connection) {
    this.emit('peer:disconnect', { peerId: connection.remotePeer })
  }

  private async _messageHandler({
    connection,
    stream,
  }: {
    connection: Connection
    stream: MuxedStream
  }) {
    try {
      await pipe(stream, async (source) => {
        for await (const msg of source) {
          const peerId = connection.remotePeer
          const message = JSON.parse(msg)

          if (
            [
              'peer:call',
              'peer:answer',
              'peer:destroy',
              'peer:announce',
              'peer:refuse',
              'peer:hangup',
              'peer:mute',
              'peer:unmute',
              'peer:screenshare',
            ].includes(message.type)
          ) {
            this.emit(message.type, { peerId, ...message })
            return
          }

          if (message.type === 'DATA') {
            this.emit('peer:data', {
              peerId,
              payload: message.payload,
            })
            return
          }

          if (message.type === 'REFUSE') {
            this.emit('peer:refuse', {
              peerId,
              payload: message.payload,
            })
            return
          }

          if (message.type === 'HANGUP') {
            this.emit('peer:hangup', {
              peerId,
              payload: message.payload,
            })
            return
          }

          if (message.type === 'TYPING_STATE') {
            this.emit('peer:typing', {
              peerId: connection.remotePeer,
              payload: message.payload,
            })
            return
          }

          if (message.type === 'SIGNAL') {
            this.emit('peer:signal', {
              peerId,
              payload: message.payload,
            })
            return
          }

          this.emit('peer:onmessage', {
            peerId,
            message,
          })
        }
      })
    } catch (error) {
      logger.log('libp2p', `Peer message error: ${error}`)
    }
  }

  private _dial(peerId: PeerId) {
    if (!this._node) return

    return this._node.dialProtocol(peerId, P2PProtocols.COMMUNICATION_BUS)
  }

  private async _sendMessage(message: string, stream: any) {
    return pipe([message], stream)
  }

  getPeer(peerId: PeerId) {
    if (!this._node) return
    return this._node.peerStore.get(peerId)
  }

  async sendMessage(message: P2PMessage, peerId: PeerId | string) {
    if (!peerId) {
      throw new Error('peerId is required')
    }
    const peer = peerId instanceof PeerId ? peerId : createFromB58String(peerId)
    if (!this._node) return

    const connection = this._node.connectionManager.get(peer)

    if (!connection) return

    const { stream } = await connection.newStream(
      P2PProtocols.COMMUNICATION_BUS,
    )

    return this._sendMessage(JSON.stringify(message), stream)
  }
}
