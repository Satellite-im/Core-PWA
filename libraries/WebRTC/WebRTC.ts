import { Config } from '~/config'
import Vue from 'vue'

import Emitter from '~/libraries/WebRTC/Emitter'
import { Wire } from '~/libraries/WebRTC/Wire'
import { WebRTCEventListeners, WireEvents } from '~/libraries/WebRTC/types'
import { Peer } from './Peer'
import { WebRTCError } from '~/store/webrtc/types'

export default class WebRTC extends Emitter<WebRTCEventListeners> {
  // Identifier to connect to signaling server with
  originator?: string
  // If this is undefined, the WebRTC services cannot run
  initialized?: boolean
  // List of peers we're actively or have been connected to
  peers: Map<string, Peer>

  // --- Internal ---
  //
  // List of functions to execute after init
  protected _fnQueue: Array<Function>
  protected _announceURLs: Array<string> = Config.webtorrent.announceURLs

  constructor() {
    super()
    this._fnQueue = []
    this.peers = new Map()
  }

  /**
   * @method init
   * @description Bind required startup data to the WebRTC class
   * @param originator identifier of the current user
   * @example
   */
  init(originator: string) {
    this.originator = originator

    this.initialized = true
    this._runQueue()
    this.emit('INIT')

    Vue.prototype.$Logger.log('WebRTC', 'Initalized', { originator: this.originator, announceURLs: this._announceURLs })
  }

  /**
   * @computed peerCount
   * @description Get the total count of connected peers
   * @returns
   * @example
   */
  get peerCount(): number {
    return this.peers ? this.peers.size : 0
  }

  // --- Private Methods ---
  //
  // Methods reserved for internal use

  /**
   * @method _queue
   * @description Queue functions that are exectued before init for execution later
   * @param fn
   */
  protected _queue(fn: Function) {
    this._fnQueue.push(fn)
  }

  /**
   * @method _runQueue
   * @description Execute all queued functions
   */
  protected _runQueue() {
    this._fnQueue.forEach((fn) => fn())
  }

  /**
   * @method setAnnounceURLs
   * @description Allow to specify different WebTorrent announce URLs for the signaling
   * @param announceURLs list of announce urls
   * @example
   */
  setAnnounceURLs(announceURLs: Array<string>) {
    this._announceURLs = announceURLs
  }

  /**
   * @method _connect
   * @description Internal abstraction of connect to allow for connection queueing
   * @param peerId identifier of peer we're connecting to
   * @param channel Secret communication channel you want to connect with
   * @returns
   * @example
   */
  protected _connect(peerId: string, channel: string): void {
    if (!this.initialized || !this.originator) {
      throw new Error(WebRTCError.NOT_INITIALIZED)
    }

    // Avoid multiple connections to the same peer
    if (this.peers.has(peerId)) {
      return
    }

    const peer = new Peer(this.originator, peerId, channel, this._announceURLs)

    this._bindWireListeners(peer.communicationBus)

    this.peers.set(peerId, peer)
  }

  protected _bindWireListeners(wire: Wire) {
    wire.on('CONNECT', ({ peerId }) => {
      this.emit('PEER_CONNECT', { peerId })
    })

    wire.on('DATA', ({ peerId, data }) => {
      Vue.prototype.$Logger.log('WebRTC', 'DATA', { peerId, data })
    })

    wire.on('ERROR', ({ peerId, error }) => {
      Vue.prototype.$Logger.log('WebRTC', 'ERROR', { peerId, error })
    })
  }

  // --- Public Methods ---
  //
  // Methods who are exposed for interaction

  /**
   * @method exists
   * @description Check if we are connected to a given peer by ID
   * @param peerId identifier of peer we're seeking
   * @returns
   * @example
   */
  exists(peerId: string): boolean {
    if (!this.peers) return false
    return this.peers.has(peerId)
  }

  /**
   * @method getPeer
   * @description Get a Wire from the list of connected peers
   * @param peerId identifier of peer we're seeking
   * @returns
   * @example
   */
  getPeer(peerId: string): Peer | undefined {
    return this.peers.get(peerId)
  }

  /**
   * @method connect
   * @description Connect to a new peer
   * @param peerId identifier of peer we're connecting to
   * @param channel Secret communication channel you want to connect with
   * @example
   */
  connect(peerId: string, channel: string) {
    if (!this.initialized) {
      this._queue(() => this._connect(peerId, channel))
    } else {
      this._connect(peerId, channel)
    }
  }
}
