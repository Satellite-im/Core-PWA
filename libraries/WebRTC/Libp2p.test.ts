import * as Libp2p from '~/libraries/WebRTC/Libp2p'

describe('Libp2p.Peer2Peer.getInstance', () => {
  test('0', () => {
    const result: any = Libp2p.Peer2Peer.getInstance()
    expect(result).toMatchSnapshot()
  })
})
