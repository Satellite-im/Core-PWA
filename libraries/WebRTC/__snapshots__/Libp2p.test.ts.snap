// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Libp2p.Peer2Peer.getInstance 0 1`] = `
Peer2Peer {
  "_events": Object {},
}
`;
